AVM-1: Logic for the API

LabiciIvan opened the pull request.

Commit:
AVM-1: Logic for the API

AndreiOrmanji commented on functions.php:
```diff
@@ -0,0 +1,272 @@
+        return null;
+    }
+
+    $products = fromJson($productsRead);
+
+    return $products ? $products : null;
```
```suggestion
    return fromJson($productsRead);
```

LabiciIvan replied:
@AndreiOrmanji - this change is fixed.

Commit:
AVM-1: Resolve readProduct() merge request comment.

AndreiOrmanji commented on functions.php:
```diff
@@ -0,0 +1,270 @@
+    try {
+        $jsonDecoded = json_decode($data, true, 512, JSON_ERROR_SYNTAX);
+    } catch (JsonException $e) {
+        echo $e->getMessage();
```
Errors should be logged in a file or in stderr. Usage of `echo` will return error to the client as a response, if it is called during request handling.

LabiciIvan replied:
@AndreiOrmanji - Function added which logs errors to the file

AndreiOrmanji commented on functions.php:
```diff
@@ -0,0 +1,270 @@
+    exit;
+}
+
+function processURL(string &$endpoint, array &$params, string &$method): void
+{
+    $url = parse_url($_SERVER['REQUEST_URI']);
+
+    $endpoint = str_replace('/', '', $url['path']);
+
+    $params = [];
+
+    if (isset($url['query'])) {
+        parse_str($url['query'], $params);
+    }
+
+    $method = $_SERVER['REQUEST_METHOD'];
+}
```
using references to set many variables is not recommended.
Instead of updating values of passed variables in this function, return their values

LabiciIvan replied:
@AndreiOrmanji - Modified this function to not set so many variables by reference and return an associative array instead.

AndreiOrmanji commented on functions.php:
```diff
@@ -0,0 +1,270 @@
+    $fields = [ID, NAME, PRICE, QUANTITY];
+
+    foreach ($fields as $property) {
+        $updatedProduct[$property] = (
+            isset($requestData[$property]) ? $requestData[$property] :
+            (isset($oldProduct[$property]) ? $oldProduct[$property] : null)
+        );
```
completely unreadable

LabiciIvan replied:
@AndreiOrmanji - Achieved a more readable syntax by using separate ternary operators to check property in each parameter.

AndreiOrmanji commented on functions.php:
```diff
@@ -0,0 +1,270 @@
+{
+    $id = 0;
+
+    if (count($categoryOfProducts) > 0) {
+        foreach ($categoryOfProducts as $product) {
+            if ($product[$idKeyName] >= $id) {
+                $id = $product[$idKeyName] + 1;
+            }
+        }
+    }
```
no need to check if array is empty before `foreach`, because if it is, body of `foreach` is unreachable, but there is no errors because of it.

LabiciIvan replied:
@AndreiOrmanji - removed the impractical use of `if` in the `generateId()` function.

AndreiOrmanji commented on functions.php:
```diff
@@ -0,0 +1,270 @@
+
+function validateRequestData(array $mandatoryFields, ?array $requestData = null): void
+{
+    if ($requestData === null) {
+        sendResponse(toJson(['error' => 'Sent data is not correct']), 414);
+    }
+
+    $missingFields = array_keys(array_diff_key($mandatoryFields, $requestData));
+
+    if ($missingFields) {
+        sendResponse(toJson(['error' => 'Missing fields: ' . implode(', ', $missingFields)]), 400);
+    }
+
+    $typeErrors = validateType($mandatoryFields, $requestData);
+
+    if($typeErrors) {
+        sendResponse(toJson(['error' => implode(', ', $typeErrors)]), 500);
+    }
```
Logic of validation, should only validate, but it is even generating response with error descriptions. Logic of validation should only generate error description, but it's formatting and outputting to client should be left to the caller of this method.

LabiciIvan replied:
@AndreiOrmanji - Adjustments added to `validateRequestData()` function which returns the error in string format.

The caller of this function will output the errors to the client.

AndreiOrmanji commented on functions.php:
```diff
@@ -0,0 +1,270 @@
+    foreach ($rules as $key => $value) {
+        switch ($value) {
+            case 'int':
+                if ($optional && isset($requestData[$key]) && !is_int($requestData[$key])) {
+                    $errors[] = "Key {$key} must be {$value}";
+                } elseif (!$optional && !is_int($requestData[$key])) {
+                    $errors[] = "Key {$key} must be {$value}";
+                }
```
```suggestion
                if ($optional && !isset($requestData[$key])) {
                    continue;
                }

                if (!$optional && !isset($requestData[$key])) {
                    $errors[] = "Key {$key} is required";
                }

                if (!is_int($requestData[$key])) {
                    $errors[] = "Key {$key} must be {$value}";
                }
```

LabiciIvan replied:
@AndreiOrmanji - I think the use of `continue` in a switch is equivalent to `break` .

As the logic of the `validateType()` is to check the type of the fields in array and in situations of when `$optional` flag is used then to check the type only if it exists in the array, I believe this change would be a better approach: 

```
                if ($optional && !isset($requestData[$key])) {
                    break;
                }

                if (!is_int($requestData[$key])) {
                    $errors[] = "Key {$key} must be {$value}";
                }
                break;
```

AndreiOrmanji replied:
@LabiciIvan, ok, use your variant here.

LabiciIvan replied:
@AndreiOrmanji - Alright, used my variant here.

AndreiOrmanji commented on functions.php:
```diff
@@ -0,0 +1,270 @@
+        }
+    }
+
+    return $errors ? $errors : null;
```
I see no difference between `[]` and `null` for case when there are no errors.

LabiciIvan replied:
@AndreiOrmanji  - Supposed to return `$errors` in case it's an empty array or use a null coalescing operator ?

LabiciIvan replied:
@AndreiOrmanji  - If there are no differences then will remove the ternary operator and return `$errors` array instead.

AndreiOrmanji replied:
`[]` can represent "no errors" case. I suggest to return array or errors in any case.

LabiciIvan replied:
@AndreiOrmanji - I used your suggestion in this case.

AndreiOrmanji reviewed (commented):
> @AndreiOrmanji - this change is fixed. 

There are other similar places to fix as well

Commit:
AVM-1: Function to log errors to the file

Commit:
AVM-1: Modified processURL() function.

- modified function to return an associative array with endpoint, params
  and method keys;

- each key from returned array is assigned to variables which use
  the corresponding key name;

Commit:
AVM-1: Removed the impractical use of if in generateId() function.

Commit:
AVM-1: Alter ternary operator in addToList() function.

- structural change in addToList() function which makes the ternary
  operator more readable;

AndreiOrmanji commented on functions.php:
```diff
@@ -0,0 +1,282 @@
+        return null;
+    }
+
+    if ($id !== null) {
+        foreach ($products[$category] as $product) {
+            if ($product['id'] == $id) {
+                return $product;
+            }
+        }
+
+        return null;
+    }
+
+    return $products[$category];
```
```suggestion
    if ($id === null) {
        return $products[$category];
    }

    foreach ($products[$category] as $product) {
        if ($product['id'] == $id) {
            return $product;
        }
    }

    return null;
```

LabiciIvan replied:
@AndreiOrmanji - Added suggested change to the `extractProduct()` function.

AndreiOrmanji commented on functions.php:
```diff
@@ -0,0 +1,282 @@
+    $fields = [ID, NAME, PRICE, QUANTITY];
+
+    foreach ($fields as $property) {
+        $isInRequest = isset($requestData[$property]) ? $requestData[$property] : null;
+        $isInOldProduct = isset($oldProduct[$property]) ? $oldProduct[$property] : null;
+
+        $updatedProduct[$property] = $isInRequest ? $isInRequest : ($isInOldProduct ?? null);
```
```suggestion
        $updatedProduct[$property] = $requestData[$property] ?? $oldProduct[$property] ?? null;
```

LabiciIvan replied:
@AndreiOrmanji  - Added this change which is a lot more easier to read then the previous change.

AndreiOrmanji commented on functions.php:
```diff
@@ -0,0 +1,282 @@
+        $updatedProduct[$property] = $isInRequest ? $isInRequest : ($isInOldProduct ?? null);
+    }
+
+    if (!isset($products[$requestData[CATEGORY]])) {
+        $products[$requestData[CATEGORY]] = [];
+    }
+
+    if ($newId) {
+        $updatedProduct[ID] = generateId($products[$requestData[CATEGORY]], ID);
+        $products[$requestData[CATEGORY]][] = $updatedProduct;
+    } else {
+        foreach($products[$requestData[CATEGORY]] as &$product) {
+            if ($product[ID] === $updatedProduct[ID]) {
+                $product = $updatedProduct;
+            }
+        }
+    }
```
```suggestion
    $category = $requestData[CATEGORY];
    if (!isset($products[$category])) {
        $products[$category] = [];
    }

    if ($newId) {
        $updatedProduct[ID] = generateId($products[$category], ID);
        $products[$category][] = $updatedProduct;
    } else {
        foreach($products[$category] as &$product) {
            if ($product[ID] === $updatedProduct[ID]) {
                $product = $updatedProduct;
            }
        }
    }
```

LabiciIvan replied:
@AndreiOrmanji - Added this suggested change.

Commit:
AVM-1: Removed usage of variables from array key.

Commit:
AVM-1: Adjustments on the validateRequestData() function.

- removed the output of validation errors within validateRequestData()
  function;

- validation errors are returned by the function and send to the user by
  the caller of this function;

Commit:
AVM-1: Removed redundant ternary operator in validateType() function.

- redundant ternary operator in validateType() function was removed and
  return $errors array which can be empty erray;

Commit:
AVM-1: Additional change for extractProduct() function.

- added the suggested change to the extractProduct() function;

Commit:
AVM-1: Suggested change to the addToList() function.

- added the suggested change to the addToList() function;

Commit:
AVM-1: Simplified validateType() function.
